Upgrade Fast DDS to v3.1.0

MiguelCompany opened the pull request:
This upgrades Fast DDS to 3.0.x

- Depends on ros2/rosidl_dynamic_typesupport_fastrtps#5
- Depends on ros2/rmw_fastrtps#776
- Allows removing `fastrtps_cmake_module`: https://github.com/ros2/rosidl_typesupport_fastrtps/pull/120

MiguelCompany commented:
CI built with `--packages-up-to rclcpp rclpy test_communication` and run with `--packages-above fastdds --packages-up-to rclcpp rclpy test_communication`:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=21804)](http://ci.ros2.org/job/ci_linux/21804/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=16075)](http://ci.ros2.org/job/ci_linux-aarch64/16075/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=1307)](http://ci.ros2.org/job/ci_linux-rhel/1307/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=22505)](http://ci.ros2.org/job/ci_windows/22505/)

clalancette commented:
I just went through the relevant PRs and did a quick compile test.  Once https://github.com/ros2/rmw_fastrtps/pull/776#pullrequestreview-2490288452 and https://github.com/ros2/rosidl_dynamic_typesupport_fastrtps/pull/5#pullrequestreview-2490286201 are fixed, we can give this a full test in CI.

MiguelCompany commented:
> I just went through the relevant PRs and did a quick compile test. Once [ros2/rmw_fastrtps#776 (review)](https://github.com/ros2/rmw_fastrtps/pull/776#pullrequestreview-2490288452) and [ros2/rosidl_dynamic_typesupport_fastrtps#5 (review)](https://github.com/ros2/rosidl_dynamic_typesupport_fastrtps/pull/5#pullrequestreview-2490286201) are fixed, we can give this a full test in CI.

@clalancette Thanks for the review. I addressed the suggestions.

Full CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=22240)](http://ci.ros2.org/job/ci_linux/22240/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=16500)](http://ci.ros2.org/job/ci_linux-aarch64/16500/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=1762)](http://ci.ros2.org/job/ci_linux-rhel/1762/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23007)](http://ci.ros2.org/job/ci_windows/23007/)

MiguelCompany commented:
> * Windows [![Build Status](https://camo.githubusercontent.com/350408fa17ccb84c6e373a2b20b53ddb49f387ab6cafb81a17792ba1c3f517af/687474703a2f2f63692e726f73322e6f72672f6275696c645374617475732f69636f6e3f6a6f623d63695f77696e646f7773266275696c643d3233303037)](http://ci.ros2.org/job/ci_windows/23007/)

This didn't fail on my local windows machine. I'm investigating the differences on the building environment.

MiguelCompany commented:
@clalancette I could reproduce the build failure on Windows (and now I have a container equivalent to the one that runs the CI).

The Windows build fails with `3.1.x` but is successful with `3.0.x` and `v3.1.0`. I have updated this PR to change Fast DDS to v3.1.0, which is currently the latest released version.

I also updated the [gist](https://gist.github.com/MiguelCompany/e22d75395d1f17b066341be8f10d0dd4/raw/9afd00801a61faedfa32a3d9101fd47aa0802ebf/ros2.repos) with the repos file. Feel free to launch CI if you agree on using v3.1.0

clalancette commented:
Bah, my mistake in CI.  I used the wrong gist.  Will fix it in a minute...

MiguelCompany commented:
@clalancette Could you check whether any of the failed tests seem related to the changes here?

clalancette commented:
So the [projectroot.test_tutorial_parameter_events__rmw_connextdds](https://ci.ros2.org/job/ci_windows/23024/testReport/junit/(root)/projectroot/test_tutorial_parameter_events__rmw_connextdds/) one is a known flake (and obviously uses Connext), so is not involved here.

However, the rest of them seem to be new.  The last couple of days have seen a lot of instability on Windows, but the 2 most recent jobs that are worth looking at are https://ci.ros2.org/view/nightly/job/nightly_win_rel/3262/ and https://ci.ros2.org/view/nightly/job/nightly_win_rel/3263/

MiguelCompany commented:
After rebasing the branches of the related PRs, I could not reproduce the Windows CI failures locally, so I just launched a new CI here:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=22379)](http://ci.ros2.org/job/ci_linux/22379/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=16642)](http://ci.ros2.org/job/ci_linux-aarch64/16642/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=1911)](http://ci.ros2.org/job/ci_linux-rhel/1911/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23172)](http://ci.ros2.org/job/ci_windows/23172/)

clalancette commented:
There is one crash on Windows which seems somewhat suspicious.  Nevertheless, I'm going to go ahead and merge all of these since I think we are pretty close here.
